File: benchmarks/cpp/oop_bench.cpp
// OOP Benchmark - C++ version for comparison with TML
// Compile: clang++ -O3 -o oop_bench_cpp oop_bench.cpp
#include <chrono>
#include <cmath>
#include <cstdint>
#include <iostream>

// ============================================================================
// 1. Virtual Dispatch Benchmark - Shape hierarchy
// ============================================================================

class Shape {
public:
    virtual ~Shape() = default;
    virtual double area() const = 0;
    virtual double perimeter() const = 0;
};

class Circle : public Shape {
    double radius;

public:
    explicit Circle(double r) : radius(r) {}
    double area() const override {
        return 3.14159265359 * radius * radius;
    }
    double perimeter() const override {
        return 2.0 * 3.14159265359 * radius;
    }
};

class Rectangle : public Shape {
    double width, height;

public:
    Rectangle(double w, double h) : width(w), height(h) {}
    double area() const override {
        return width * height;
    }
    double perimeter() const override {
        return 2.0 * (width + height);
    }
};

class Triangle : public Shape {
    double a, b, c, h;

public:
    Triangle(double a_, double b_, double c_, double h_) : a(a_), b(b_), c(c_), h(h_) {}
    double area() const override {
        return 0.5 * a * h;
    }
    double perimeter() const override {
        return a + b + c;
    }
};

double virtual_dispatch_bench(int iterations) {
    Circle circle(5.0);
    Rectangle rect(4.0, 6.0);
    Triangle tri(3.0, 4.0, 5.0, 5.0);

    double total = 0.0;
    for (int i = 0; i < iterations; ++i) {
        int mod = i % 3;
        if (mod == 0) {
            total += circle.area() + circle.perimeter();
        } else if (mod == 1) {
            total += rect.area() + rect.perimeter();
        } else {
            total += tri.area() + tri.perimeter();
        }
    }
    return total;
}

// ============================================================================
// 2. Object Creation Benchmark
// ============================================================================

class Point {
    double x, y;

public:
    Point(double x_, double y_) : x(x_), y(y_) {}
    double distance(const Point& other) const {
        double dx = x - other.x;
        double dy = y - other.y;
        return std::sqrt(dx * dx + dy * dy);
    }
};

double object_creation_bench(int iterations) {
    double total = 0.0;
    for (int i = 0; i < iterations; ++i) {
        double fi = static_cast<double>(i);
        Point p1(fi, fi * 2.0);
        Point p2(fi + 1.0, fi + 3.0);
        total += p1.distance(p2);
    }
    return total;
}

// ============================================================================
// 3. HTTP Handler Benchmark
// ============================================================================

class HttpHandler {
public:
    virtual ~HttpHandler() = default;
    virtual int handle(int method, int path_id) = 0;
};

class GetHandler : public HttpHandler {
    int resource_id;

public:
    explicit GetHandler(int id) : resource_id(id) {}
    int handle(int method, int path_id) override {
        return resource_id + method + path_id;
    }
};

class PostHandler : public HttpHandler {
    int resource_id;
    bool validate;

public:
    PostHandler(int id, bool v) : resource_id(id), validate(v) {}
    int handle(int method, int path_id) override {
        int result = resource_id + method + path_id;
        if (validate)
            result += 100;
        return result;
    }
};

class DeleteHandler : public HttpHandler {
    int resource_id;

public:
    explicit DeleteHandler(int id) : resource_id(id) {}
    int handle(int method, int path_id) override {
        return resource_id - method + path_id;
    }
};

int http_handler_bench(int iterations) {
    GetHandler get_h(1);
    PostHandler post_h(2, true);
    DeleteHandler del_h(3);

    int total = 0;
    for (int i = 0; i < iterations; ++i) {
        int mod = i % 3;
        if (mod == 0) {
            total += get_h.handle(1, 0);
        } else if (mod == 1) {
            total += post_h.handle(2, 1);
        } else {
            total += del_h.handle(3, 2);
        }
    }
    return total;
}

// ============================================================================
// 4. Game Loop Benchmark
// ============================================================================

class GameObject {
public:
    virtual ~GameObject() = default;
    virtual double update(double dt) = 0;
};

class Player : public GameObject {
    double x, y, speed;

public:
    Player(double x_, double y_, double s) : x(x_), y(y_), speed(s) {}
    double update(double dt) override {
        x += speed * dt;
        y += speed * dt * 0.5;
        return x + y;
    }
};

class Enemy : public GameObject {
    double x, y, speed, target_x, target_y;

public:
    Enemy(double x_, double y_, double s, double tx, double ty)
        : x(x_), y(y_), speed(s), target_x(tx), target_y(ty) {}
    double update(double dt) override {
        double dx = target_x - x;
        double dy = target_y - y;
        double dist = std::sqrt(dx * dx + dy * dy);
        if (dist > 0.001) {
            x += (dx / dist) * speed * dt;
            y += (dy / dist) * speed * dt;
        }
        return x + y;
    }
};

class Projectile : public GameObject {
    double x, y, vx, vy;

public:
    Projectile(double x_, double y_, double vx_, double vy_) : x(x_), y(y_), vx(vx_), vy(vy_) {}
    double update(double dt) override {
        x += vx * dt;
        y += vy * dt;
        return x + y;
    }
};

double game_loop_bench(int iterations) {
    Player player(0.0, 0.0, 5.0);
    Enemy enemy(100.0, 100.0, 3.0, 0.0, 0.0);
    Projectile proj(0.0, 0.0, 10.0, 10.0);

    double total = 0.0;
    double dt = 0.016;
    for (int i = 0; i < iterations; ++i) {
        total += player.update(dt);
        total += enemy.update(dt);
        total += proj.update(dt);
    }
    return total;
}

// ============================================================================
// 5. Deep Inheritance Benchmark
// ============================================================================

class Base {
protected:
    int a;

public:
    explicit Base(int a_) : a(a_) {}
    virtual ~Base() = default;
    virtual int compute() {
        return a;
    }
};

class Derived1 : public Base {
protected:
    int b;

public:
    Derived1(int a_, int b_) : Base(a_), b(b_) {}
    int compute() override {
        return Base::compute() + b;
    }
};

class Derived2 : public Derived1 {
protected:
    int c;

public:
    Derived2(int a_, int b_, int c_) : Derived1(a_, b_), c(c_) {}
    int compute() override {
        return Derived1::compute() + c;
    }
};

class Derived3 : public Derived2 {
protected:
    int d;

public:
    Derived3(int a_, int b_, int c_, int d_) : Derived2(a_, b_, c_), d(d_) {}
    int compute() override {
        return Derived2::compute() + d;
    }
};

class Derived4 : public Derived3 {
protected:
    int e;

public:
    Derived4(int a_, int b_, int c_, int d_, int e_) : Derived3(a_, b_, c_, d_), e(e_) {}
    int compute() override {
        return Derived3::compute() + e;
    }
};

int deep_inheritance_bench(int iterations) {
    Base base(10);
    Derived1 d1(10, 5);
    Derived2 d2(10, 5, 3);
    Derived3 d3(10, 5, 3, 2);
    Derived4 d4(10, 5, 3, 2, 1);

    int total = 0;
    for (int i = 0; i < iterations; ++i) {
        int mod = i % 5;
        if (mod == 0)
            total += base.compute();
        else if (mod == 1)
            total += d1.compute();
        else if (mod == 2)
            total += d2.compute();
        else if (mod == 3)
            total += d3.compute();
        else
            total += d4.compute();
    }
    return total;
}

// ============================================================================
// 6. Method Chaining Benchmark
// ============================================================================

class Builder {
    int val;

public:
    Builder() : val(0) {}
    Builder with_add(int n) {
        Builder b;
        b.val = val + n;
        return b;
    }
    Builder with_multiply(int n) {
        Builder b;
        b.val = val * n;
        return b;
    }
    Builder with_subtract(int n) {
        Builder b;
        b.val = val - n;
        return b;
    }
    int build() const {
        return val;
    }
};

int method_chaining_bench(int iterations) {
    int total = 0;
    for (int i = 0; i < iterations; ++i) {
        int result = Builder().with_add(10).with_multiply(2).with_subtract(5).with_add(i).build();
        total += result;
    }
    return total;
}

// ============================================================================
// Main
// ============================================================================

int64_t time_ns() {
    using namespace std::chrono;
    return duration_cast<nanoseconds>(high_resolution_clock::now().time_since_epoch()).count();
}

// Prevent compiler from optimizing away results
volatile double sink_double = 0;
volatile int sink_int = 0;

template <typename F> int64_t run_benchmark_d(F func, int iterations) {
    int warmup = iterations / 10;
    sink_double = func(warmup);

    int64_t start = time_ns();
    sink_double = func(iterations);
    return time_ns() - start;
}

template <typename F> int64_t run_benchmark_i(F func, int iterations) {
    int warmup = iterations / 10;
    sink_int = func(warmup);

    int64_t start = time_ns();
    sink_int = func(iterations);
    return time_ns() - start;
}

int main() {
    std::cout << "============================================\n";
    std::cout << "C++ OOP Performance Benchmarks\n";
    std::cout << "============================================\n\n";

    int iterations = 100000;
    std::cout << "Running " << iterations << " iterations per benchmark...\n\n";

    int64_t t1 = run_benchmark_d(virtual_dispatch_bench, iterations);
    int64_t t2 = run_benchmark_d(object_creation_bench, iterations);
    int64_t t3 = run_benchmark_i(http_handler_bench, iterations);
    int64_t t4 = run_benchmark_d(game_loop_bench, iterations);
    int64_t t5 = run_benchmark_i(deep_inheritance_bench, iterations);
    int64_t t6 = run_benchmark_i(method_chaining_bench, iterations);

    std::cout << "Virtual Dispatch:     " << t1 / 1000 << " us\n";
    std::cout << "Object Creation:      " << t2 / 1000 << " us\n";
    std::cout << "HTTP Handler:         " << t3 / 1000 << " us\n";
    std::cout << "Game Loop:            " << t4 / 1000 << " us\n";
    std::cout << "Deep Inheritance:     " << t5 / 1000 << " us\n";
    std::cout << "Method Chaining:      " << t6 / 1000 << " us\n";

    std::cout << "\n============================================\n";
    std::cout << "Benchmarks complete!\n";
    std::cout << "============================================\n";

    return 0;
}

File: benchmarks/rust/oop_bench.rs
// OOP Performance Benchmark for Rust
// Equivalent to oop_bench.cpp and oop_bench.tml
use std::time::Instant;

// =============================================================================
// Benchmark 1: Virtual Dispatch (trait objects)
// =============================================================================

trait Shape {
    fn area(&self) -> f64;
    fn perimeter(&self) -> f64;
}

struct Circle {
    radius: f64,
}

impl Shape for Circle {
    fn area(&self) -> f64 {
        std::f64::consts::PI * self.radius * self.radius
    }
    fn perimeter(&self) -> f64 {
        2.0 * std::f64::consts::PI * self.radius
    }
}

struct Rectangle {
    width: f64,
    height: f64,
}

impl Shape for Rectangle {
    fn area(&self) -> f64 {
        self.width * self.height
    }
    fn perimeter(&self) -> f64 {
        2.0 * (self.width + self.height)
    }
}

struct Triangle {
    base: f64,
    height: f64,
    side1: f64,
    side2: f64,
}

impl Shape for Triangle {
    fn area(&self) -> f64 {
        0.5 * self.base * self.height
    }
    fn perimeter(&self) -> f64 {
        self.base + self.side1 + self.side2
    }
}

fn virtual_dispatch_bench(iterations: i32) -> f64 {
    let circle = Circle { radius: 5.0 };
    let rectangle = Rectangle { width: 4.0, height: 6.0 };
    let triangle = Triangle { base: 3.0, height: 4.0, side1: 5.0, side2: 5.0 };

    let mut result = 0.0;
    for i in 0..iterations {
        let shape: &dyn Shape = match i % 3 {
            0 => &circle,
            1 => &rectangle,
            _ => &triangle,
        };
        result += shape.area();
        result += shape.perimeter();
    }
    result
}

// =============================================================================
// Benchmark 2: Object Creation
// =============================================================================

#[derive(Clone)]
struct Point {
    x: f64,
    y: f64,
}

impl Point {
    fn new(x: f64, y: f64) -> Self {
        Point { x, y }
    }

    fn distance(&self, other: &Point) -> f64 {
        let dx = self.x - other.x;
        let dy = self.y - other.y;
        (dx * dx + dy * dy).sqrt()
    }
}

fn object_creation_bench(iterations: i32) -> f64 {
    let mut result = 0.0;
    for i in 0..iterations {
        let p1 = Point::new(i as f64, i as f64 * 2.0);
        let p2 = Point::new(i as f64 + 1.0, i as f64 + 3.0);
        result += p1.distance(&p2);
    }
    result
}

// =============================================================================
// Benchmark 3: HTTP Handler Pattern (trait objects)
// =============================================================================

trait HttpHandler {
    fn handle(&self, method: i32, path: i32) -> i32;
}

struct GetHandler {
    route_id: i32,
}

impl HttpHandler for GetHandler {
    fn handle(&self, _method: i32, path: i32) -> i32 {
        self.route_id * 100 + path
    }
}

struct PostHandler {
    route_id: i32,
    requires_auth: bool,
}

impl HttpHandler for PostHandler {
    fn handle(&self, method: i32, path: i32) -> i32 {
        let auth_bonus = if self.requires_auth { 1000 } else { 0 };
        self.route_id * 100 + path + method + auth_bonus
    }
}

struct DeleteHandler {
    route_id: i32,
}

impl HttpHandler for DeleteHandler {
    fn handle(&self, method: i32, path: i32) -> i32 {
        self.route_id * 100 + path - method
    }
}

fn http_handler_bench(iterations: i32) -> i32 {
    let get_handler = GetHandler { route_id: 1 };
    let post_handler = PostHandler { route_id: 2, requires_auth: true };
    let delete_handler = DeleteHandler { route_id: 3 };

    let mut result = 0;
    for i in 0..iterations {
        let handler: &dyn HttpHandler = match i % 3 {
            0 => &get_handler,
            1 => &post_handler,
            _ => &delete_handler,
        };
        result += handler.handle(i % 4, i % 100);
    }
    result
}

// =============================================================================
// Benchmark 4: Game Loop Pattern (trait objects with state)
// =============================================================================

trait Entity {
    fn update(&mut self, delta: f64);
    fn get_x(&self) -> f64;
    fn get_y(&self) -> f64;
}

struct Player {
    x: f64,
    y: f64,
    speed: f64,
    health: i32,
}

impl Entity for Player {
    fn update(&mut self, delta: f64) {
        self.x += self.speed * delta;
        self.y += self.speed * delta * 0.5;
    }
    fn get_x(&self) -> f64 { self.x }
    fn get_y(&self) -> f64 { self.y }
}

struct Enemy {
    x: f64,
    y: f64,
    speed: f64,
    damage: i32,
}

impl Entity for Enemy {
    fn update(&mut self, delta: f64) {
        self.x -= self.speed * delta;
        self.y += (self.x * 0.1).sin() * delta;
    }
    fn get_x(&self) -> f64 { self.x }
    fn get_y(&self) -> f64 { self.y }
}

struct Projectile {
    x: f64,
    y: f64,
    vx: f64,
    vy: f64,
}

impl Entity for Projectile {
    fn update(&mut self, delta: f64) {
        self.x += self.vx * delta;
        self.y += self.vy * delta;
    }
    fn get_x(&self) -> f64 { self.x }
    fn get_y(&self) -> f64 { self.y }
}

fn game_loop_bench(iterations: i32) -> f64 {
    let mut player = Player { x: 0.0, y: 0.0, speed: 10.0, health: 100 };
    let mut enemy = Enemy { x: 100.0, y: 50.0, speed: 5.0, damage: 10 };
    let mut projectile = Projectile { x: 0.0, y: 0.0, vx: 20.0, vy: 15.0 };

    let delta = 0.016; // 60 FPS
    let mut result = 0.0;

    for i in 0..iterations {
        let entity: &mut dyn Entity = match i % 3 {
            0 => &mut player,
            1 => &mut enemy,
            _ => &mut projectile,
        };
        entity.update(delta);
        result += entity.get_x() + entity.get_y();
    }
    result
}

// =============================================================================
// Benchmark 5: Deep Inheritance (simulated with composition)
// =============================================================================

struct Base {
    value: i32,
}

impl Base {
    fn compute(&self) -> i32 {
        self.value
    }
}

struct Derived1 {
    value: i32,
    bonus1: i32,
}

impl Derived1 {
    fn compute(&self) -> i32 {
        self.value + self.bonus1
    }
}

struct Derived2 {
    value: i32,
    bonus1: i32,
    bonus2: i32,
}

impl Derived2 {
    fn compute(&self) -> i32 {
        self.value + self.bonus1 + self.bonus2
    }
}

struct Derived3 {
    value: i32,
    bonus1: i32,
    bonus2: i32,
    bonus3: i32,
}

impl Derived3 {
    fn compute(&self) -> i32 {
        self.value + self.bonus1 + self.bonus2 + self.bonus3
    }
}

struct Derived4 {
    value: i32,
    bonus1: i32,
    bonus2: i32,
    bonus3: i32,
    bonus4: i32,
}

impl Derived4 {
    fn compute(&self) -> i32 {
        self.value + self.bonus1 + self.bonus2 + self.bonus3 + self.bonus4
    }
}

fn deep_inheritance_bench(iterations: i32) -> i32 {
    let base = Base { value: 10 };
    let d1 = Derived1 { value: 10, bonus1: 5 };
    let d2 = Derived2 { value: 10, bonus1: 5, bonus2: 3 };
    let d3 = Derived3 { value: 10, bonus1: 5, bonus2: 3, bonus3: 2 };
    let d4 = Derived4 { value: 10, bonus1: 5, bonus2: 3, bonus3: 2, bonus4: 1 };

    let mut result = 0;
    for i in 0..iterations {
        result += match i % 5 {
            0 => base.compute(),
            1 => d1.compute(),
            2 => d2.compute(),
            3 => d3.compute(),
            _ => d4.compute(),
        };
    }
    result
}

// =============================================================================
// Benchmark 6: Method Chaining / Builder Pattern
// =============================================================================

#[derive(Clone, Default)]
struct Builder {
    value: i32,
}

impl Builder {
    fn new() -> Self {
        Builder { value: 0 }
    }

    fn add(mut self, n: i32) -> Self {
        self.value += n;
        self
    }

    fn multiply(mut self, n: i32) -> Self {
        self.value *= n;
        self
    }

    fn subtract(mut self, n: i32) -> Self {
        self.value -= n;
        self
    }

    fn build(self) -> i32 {
        self.value
    }
}

fn method_chaining_bench(iterations: i32) -> i32 {
    let mut result = 0;
    for i in 0..iterations {
        let value = Builder::new()
            .add(i)
            .multiply(2)
            .subtract(1)
            .add(i % 10)
            .multiply(3)
            .build();
        result += value;
    }
    result
}

// =============================================================================
// Benchmark Runner
// =============================================================================

fn run_benchmark<F, R>(name: &str, iterations: i32, f: F) -> u64
where
    F: Fn(i32) -> R,
{
    // Warmup
    let warmup = iterations / 10;
    let _ = f(warmup);

    // Timed run
    let start = Instant::now();
    let _ = f(iterations);
    let elapsed = start.elapsed();

    elapsed.as_micros() as u64
}

fn main() {
    println!("============================================");
    println!("Rust OOP Performance Benchmarks");
    println!("============================================");
    println!();

    let iterations = 100000;
    println!("Running {} iterations per benchmark...", iterations);
    println!();

    let t1 = run_benchmark("Virtual Dispatch", iterations, virtual_dispatch_bench);
    let t2 = run_benchmark("Object Creation", iterations, object_creation_bench);
    let t3 = run_benchmark("HTTP Handler", iterations, http_handler_bench);
    let t4 = run_benchmark("Game Loop", iterations, game_loop_bench);
    let t5 = run_benchmark("Deep Inheritance", iterations, deep_inheritance_bench);
    let t6 = run_benchmark("Method Chaining", iterations, method_chaining_bench);

    println!("Virtual Dispatch:     {} us", t1);
    println!("Object Creation:      {} us", t2);
    println!("HTTP Handler:         {} us", t3);
    println!("Game Loop:            {} us", t4);
    println!("Deep Inheritance:     {} us", t5);
    println!("Method Chaining:      {} us", t6);

    println!();
    println!("============================================");
    println!("Benchmarks complete!");
    println!("============================================");
}
